saturn_client/core.py

```diff
@@ -0,0 +1,171 @@
+    ) -> Dict[str, Any]:
+        """
+        Create a project from scratch
+
+        :param name: Name of project.
+        :param description: Short description of the project (less than 250 characters).
+        :param image_uri: Location of the image. Example:
+            485185227295.dkr.ecr.us-east-1.amazonaws.com/saturn-dask:2020.12.01.21.10
+        :param start_script: Script that runs on start up. Examples: "pip install dask"
+        :param environment_variables: Env vars expressed as a dict. The names will be
+            coerced to uppercase.
+        :param working_dir: Location to use as working directory. Example: /home/jovyan/project
+        :param workspace_settings: Setting for the jupyter associated with the project.
+            These include: `size`, `disk_space`, `auto_shutoff`, and `start_ssh`. The options
+            for these are available from `conn.options`.
+        """
+        url = urljoin(self.url, "api/projects")
+
+        errors = []
+        if environment_variables:
+            environment_variables = "\n".join(
+                f"{k.upper()}={v}" for k, v in environment_variables.items()
+            )
+
+        workspace_kwargs = {}
+        if workspace_settings:
+            for k, v in workspace_settings.items():
+                workspace_keys = ["size", "auto_shutoff", "disk_space", "start_ssh"]
+                if k in workspace_keys:
+                    if k == "start_ssh":
+                        if not isinstance(v, bool):
+                            errors.append(f"{k} must be set to a boolean if defined.")
+                    else:
+                        options = (
+                            list(self.options["sizes"].keys()) if k == "size" else self.options[k]
+                        )
+                        if v not in options:
+                            errors.append(
+                                f"Proposed {k}: {v} is not a valid option. Options are: {options}."
+                            )
+                    workspace_kwargs[f"jupyter_{k}"] = v
+                else:
+                    errors.append(
+                        f"{k} is not a valid workspace_setting. "
+                        "Supported workspace_settings are {workspace_keys}."
+                    )
+
+        if len(errors) > 0:
+            raise ValueError(errors)
+
+        project_config = {
+            "name": name,
+            "description": description,
+            "image": image_uri,
+            "start_script": start_script,
+            "environment_variables": environment_variables,
+            "working_dir": working_dir,
+            **workspace_kwargs,
+        }
+        # only send kwargs that are explicitly set by user
+        project_config = {k: v for k, v in project_config.items() if v is not None}
+
+        response = requests.post(
+            url,
+            data=json.dumps({**project_config, "saturn_client_version": __version__}),
```

Comment:
Why part of the request body? Wouldn't a custom header (like `X-Saturn-Client-Version`) be less intrusive? Or does this route specifically use this field?

Reply:
I thought the route should be able to take different actions or serve different errors based on the version. I have been thinking about this pattern for dask-saturn as well. I have no pre-set notions of how people normally pass this type of information around.

Reply:
Via custom header is how I've seen it done before. I _believe_ the AWS CLI uses headers, but I hesitate to dig through that codebase to confirm ;)

Reply:
good call

Reply:
Changed :) https://github.com/saturncloud/saturn/pull/1572 for the saturn side